camds_agent.py

```diff
@@ -0,0 +1,278 @@
+from packet import Packet
+from common import PacketType
+
+
+# Camera will never have more than 1000 images
+MAX_NUM_FILES = 1000
+
+DOWNLOAD_TIMEOUT = 120
+
+IMG_DIR_ROOT = '/home/asadev/camds_images/'
+
+IMG_REGEX = '.*(\d{4})(\d{2})(\d{2})T.*jpg'
+IMG_PATTERN = re.compile(IMG_REGEX, re.DOTALL)
+
+
+class CamdsPortAgent(TcpPortAgent):
+
+	img_dir_root = IMG_DIR_ROOT
+
+	def __init__(self, config):
+		super(CamdsPortAgent, self).__init__(config)
+
+		self.img_dir_root = config['imgdir']
+
+		self._start_smb_connection()
+
+		log.msg('CamdsPortAgent initialization complete')
+
+	def my_callback(self):
+		pass
+
+	def _start_smb_connection(self):
+
+		# Credentials to login to the camera
+		userID = 'subc'
+		password = 'password'
+		share_name = 'Media'
+
+		# client_machine_name can be an arbitary ASCII string
+		# server_name should match the remote machine name, or else the connection will be rejected
+		client_machine_name = 'dummy_machine'
+		server_name = self.getServerName(self.inst_addr)[0]
+
+		log.msg('Server name: ', server_name)
+
+		download_factory = RetrieveFileFactory(self.inst_addr, self.router, self.refdes, self.img_dir_root, share_name,
+											   userID, password, client_machine_name, server_name, use_ntlm_v2=True)
+
+		reactor.connectTCP(host=self.inst_addr, port=139, factory=download_factory)
+
+	def getServerName(self, remote_smb_ip, timeout=30):
+		bios = NetBIOS()
+		srv_name = None
+		try:
+			srv_name = bios.queryIPForName(ip=remote_smb_ip, timeout=timeout)
+		except:
+			log.err("Couldn't find SMB server name, check remote_smb_ip again!!")
+		finally:
+			bios.close()
+			return srv_name
+
+# This change can be removed once Pete's fix has been published upstream
+# Original code was using iteritems() vs. items, which was causing an exception.
+# We wouldn't then need to override this method
+class FixedProtocol(SMBProtocol):
+	def _cleanupPendingRequests(self):
+
+		if self.factory.instance == self:
+			now = time.time()
+			for mid, r in self.pending_requests.items():
+				if r.expiry_time < now:
+					try:
+						r.errback(SMBTimeout())
+					except Exception as e:
+						log.err('Exception occurred while cleaning up pending request: ', str(e.message))
+
+					del self.pending_requests[mid]
+
+			reactor.callLater(1, self._cleanupPendingRequests)
+
+
+class RetrieveFileFactory(SMBProtocolFactory):
+	protocol = FixedProtocol
+
+	def __init__(self, inst_addr, router, ref_des, img_dir_root, share_name, *args, **kwargs):
+		SMBProtocolFactory.__init__(self, *args, **kwargs)
+		self.d = defer.Deferred()
+		self.instrument_ip = inst_addr
+		self.router = router
+		self.ref_des = ref_des
+		self.share_name = share_name
+
+		self.img_dir_root = img_dir_root
+		self.image_dir = self.img_dir_root + self.ref_des
+		if not os.path.exists(self.image_dir):
+			os.makedirs(self.image_dir)
+
+		self.pending_file_queue = collections.deque('')
+		self.visited_dir_queue = collections.deque('') # kcr
```

Comment:
What does this do?

Reply:
Woops. That last line is a debug line that I forgot to remove. I'm surprised that doesn't cause an error.

Reply:
Line 114 in the above comment was leftover debugging code and has removed